src/main/java/org/purescript/psi/PSModule.kt

```diff
@@ -36,16 +38,41 @@ class PSModule(node: ASTNode) :
     }
 
     /**
-     * If the export list is null, this module implicitly exports all its members.
+     * Helper method for retrieving various types of exported declarations.
+     *
+     * @param declarations The declarations of the wanted type in this module
+     * @param importedDeclarationProperty The property for the imported declarations in an [PSImportDeclarationImpl]
+     * @param exportedItemClass The class of the [PSExportedItem] to use when filtering the results
      */
-    val exportList: PSExportList? = findChildByClass(PSExportList::class.java)
+    private fun <Declaration : PsiNamedElement> getExportedDeclarations(
+        declarations: Array<Declaration>,
+        importedDeclarationProperty: KProperty1<PSImportDeclarationImpl, List<Declaration>>,
+        exportedItemClass: Class<out PSExportedItem>
+    ): List<Declaration> {
+        val explicitlyExportedItems = exportList?.exportedItems
+            ?: return declarations.toList()
+
+        val explicitlyNames = explicitlyExportedItems
+            .filterIsInstance(exportedItemClass)
+            .map { it.name }
+            .toSet()
+
+        val exportedDeclarations = mutableListOf<Declaration>()
+        declarations.filterTo(exportedDeclarations) {
+            it.name in explicitlyNames
+        }
+
+        explicitlyExportedItems.filterIsInstance<PSExportedModule>()
+            .mapNotNull { it.importDeclaration }
+            .flatMapTo(exportedDeclarations) { importedDeclarationProperty.get(it) }
+
+        return exportedDeclarations
+    }
```

Comment:
I made this helper method both here and in PSImportDeclarationImpl. If you don't like them, we can do something else. For now, it reduces the amount of code duplication.